connectors/sync_job_runner.py

```diff
@@ -0,0 +1,176 @@
+            await self.elastic_server.prepare_content_index(
+                self.sync_job.index_name, mappings=mappings
+            )
+
+            # allows the data provider to change the bulk options
+            bulk_options = self.bulk_options.copy()
+            self.data_provider.tweak_bulk_options(bulk_options)
+
+            result = await self.elastic_server.async_bulk(
+                self.sync_job.index_name,
+                self.prepare_docs(),
+                self.sync_job.pipeline,
+                filter_=self.sync_job.filtering,
+                sync_rules_enabled=sync_rules_enabled,
+                options=bulk_options,
+            )
+            sync_error = result.get("fetch_error")
+            sync_status = JobStatus.COMPLETED if sync_error is None else JobStatus.ERROR
+            await self._sync_done(
+                sync_status=sync_status, result=result, sync_error=sync_error
+            )
+        except asyncio.CancelledError:
+            await self._sync_done(sync_status=JobStatus.SUSPENDED, result={})
+        except Exception as e:
+            await self._sync_done(sync_status=JobStatus.ERROR, result={}, sync_error=e)
+        finally:
+            if self.data_provider is not None:
+                await self.data_provider.close()
+
+    async def _sync_done(self, sync_status, result=None, sync_error=None):
+        if result is None:
+            result = {}
+        doc_updated = result.get("doc_updated", 0)
+        doc_created = result.get("doc_created", 0)
+        doc_deleted = result.get("doc_deleted", 0)
+        indexed_count = doc_updated + doc_created
+
+        ingestion_stats = {
+            "indexed_document_count": indexed_count,
+            "indexed_document_volume": 0,
+            "deleted_document_count": doc_deleted,
+            "total_document_count": await self.connector.document_count(),
+        }
+
+        if sync_status == JobStatus.ERROR:
+            await self.sync_job.fail(sync_error, ingestion_stats=ingestion_stats)
+        elif sync_status == JobStatus.SUSPENDED:
+            await self.sync_job.suspend(ingestion_stats=ingestion_stats)
+        elif sync_status == JobStatus.CANCELED:
+            await self.sync_job.cancel(ingestion_stats=ingestion_stats)
+        else:
+            await self.sync_job.done(ingestion_stats=ingestion_stats)
+
+        self.sync_job = await self.sync_job.reload()
+        await self.connector.sync_done(self.sync_job)
+        logger.info(
+            f"[{self.job_id}] Sync done: {indexed_count} indexed, {doc_deleted} "
+            f" deleted. ({int(time.time() - self._start_time)} seconds)"  # pyright: ignore
+        )
+
+    async def _claim_job(self):
+        try:
+            await self.sync_job.claim()
+            await self.connector.sync_starts()
+            self._start_time = time.time()
+            return True
+        except Exception as e:
+            logger.critical(e, exc_info=True)
```

Comment:
Nit: since we recover from this error here, do we want to issue a `warning` instead?

Reply:
Actually, we don't recover from here. If an exception is caught, `_claim_job` will return `False`, and the caller will raise `JobClaimError`, so `sync_job_runner.execute` is done.

The current implementation does not take race conditions into account. 

Once we decouple the job execution to a new `JobExecutionService`, there can be multiple instances trying to claim a single job, and it's expected that claim can fail. In that case, we will just log a message, and exit without any exception.